mlflow/torchscript.py

```diff
@@ -0,0 +1,313 @@
+                          - ``relative/path/to/local/model``
+                          - ``s3://my_bucket/path/to/model``
+                          - ``runs:/<mlflow_run_id>/run-relative/path/to/model``
+                          - ``models:/<model_name>/<model_version>``
+                          - ``models:/<model_name>/<stage>``
+
+                          For more information about supported URI schemes, see
+                          `Referencing Artifacts <https://www.mlflow.org/docs/latest/concepts.html#
+                          artifact-locations>`_.
+
+        :param kwargs: kwargs to pass to ``torch.jit.load`` method.
+        :return: A torchscript model.
+
+        .. code-block:: python
+            :caption: Example
+
+            import torch
+            import mlflow
+            import mlflow.torchscript
+            # Set values
+            model_path_dir = ...
+            run_id = "96771d893a5e46159d9f3b49bf9013e2"
+            model = mlflow.torchscript.load_model("runs:/" + run_id + "/" + model_path_dir)
+            y_pred = model(x_new_data)
+        """
+    local_model_path = Path(_download_artifact_from_uri(artifact_uri=model_uri))
+    torchscript_conf = _get_flavor_configuration(model_path=local_model_path,
+                                                 flavor_name=FLAVOR_NAME)
+    if torch.__version__ != torchscript_conf["pytorch_version"]:
+        _logger.warning(
+            "Stored model version '%s' does not match installed PyTorch version '%s'",
+            torchscript_conf["pytorch_version"], torch.__version__)
+    return torch.jit.load(
+        str(local_model_path / _SERIALIZED_TORCH_MODEL_FILE_NAME), **kwargs)
+
+
+def _load_pyfunc(path, **kwargs):
+    """
+    Load PyFunc implementation. Called by ``pyfunc.load_pyfunc``.
+
+    :param path: Local filesystem path to the MLflow Model with the ``torchscript`` flavor.
+    """
+    loaded = load_model(path, **kwargs)
+    return _PyTorchWrapper(loaded)
+
+
+class _PyTorchWrapper(object):
+    """
+    Wrapper class that creates a predict function such that
+    predict(data: pd.DataFrame) -> model's output as pd.DataFrame (pandas DataFrame)
+    """
+    def __init__(self, torchscript_model):
+        self.torchscript_model = torchscript_model
+
+    def predict(self, data, device='cpu'):
+        if not isinstance(data, pd.DataFrame):
+            raise TypeError("Input data should be pandas.DataFrame")
+        self.torchscript_model.to(device)
+        self.torchscript_model.eval()
+        with torch.no_grad():
+            input_tensor = torch.from_numpy(data.values.astype(np.float32)).to(device)
+            preds = self.torchscript_model(input_tensor)
+            if not isinstance(preds, torch.Tensor):
+                raise TypeError("Expected PyTorch model to output a single output tensor, "
+                                "but got output of type '{}'".format(type(preds)))
+            predicted = pd.DataFrame(preds.numpy())
```

Comment:
BTW, what do you think about this output format? It matches what the Pytorch flavor produces, but is a bit odd - we assume that the output tensor is one-or-two-dimensional and create a DataFrame where individual elements of each tensor are put into separate columns:

```
>>> oned_tensor
array([1, 2, 3])
>>> twod_tensor
array([[1, 2, 3],
       [4, 5, 6]])
>>> pd.DataFrame(oned_tensor)
   0
0  1
1  2
2  3
>>> pd.DataFrame(twod_tensor
... )
   0  1  2
0  1  2  3
1  4  5  6
>>> threed
array([[[1, 2, 3],
        [4, 5, 6]],

       [[1, 2, 3],
        [4, 5, 6]]])
>>> pd.DataFrame(threed)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sid.murching/miniconda3/envs/mlflow-python3/lib/python3.6/site-packages/pandas/core/frame.py", line 361, in __init__
    copy=copy)
  File "/Users/sid.murching/miniconda3/envs/mlflow-python3/lib/python3.6/site-packages/pandas/core/frame.py", line 513, in _init_ndarray
    values = _prep_ndarray(values, copy=copy)
  File "/Users/sid.murching/miniconda3/envs/mlflow-python3/lib/python3.6/site-packages/pandas/core/frame.py", line 6257, in _prep_ndarray
    raise ValueError('Must pass 2-d input')
ValueError: Must pass 2-d input
```

Packaging the output into a DataFrame has the advantage that we can use it with MLflow's ``mlflow.pyfunc.spark_udf`` API, which expects the pyfunc to produce DataFrame-able output, but I wonder if we should just return a numpy array here to avoid constraining what the torch model can produce as output.

Reply:
On the other hand, this approach is consistent with `mlflow.pytorch` and we could argue models usually produce 1D output (for regression) or 2D output (for classification when outputting probabilities across classes), in which case this approach is fine. Let me ask some folks more familiar with Pytorch use cases for opinions - currently I'm thinking this is fine as is, but if we do e.g. extend MLflow's spark UDF to better support multidimensional outputs, maybe we should consider the numpy array approach

Reply:
Right, It is similar to "allowing users to pass only one input" scenario. My experience is mostly working with DL algorithms and hence my viewpoint is that restricting any of the DL frameworks to output DF objects is not nice. But then, that needs to be changed across the MLflow project. Let me know your decision.